File: templates/Uwp/_composition/_shared/Page.Ink.AddUndoRedoServiceVB/Services/Ink/InkUndoRedoService.vb
Imports System.Collections.Generic
Imports System.Linq
Imports Windows.UI.Xaml.Controls
Imports Param_ItemNamespace.EventHandlers.Ink
Imports Param_ItemNamespace.Services.Ink.UndoRedo

Namespace Services.Ink
    Public Class InkUndoRedoService
        Private ReadOnly _strokeService As InkStrokesService
        Private undoStack As Stack(Of IUndoRedoOperation) = New Stack(Of IUndoRedoOperation)()
        Private redoStack As Stack(Of IUndoRedoOperation) = New Stack(Of IUndoRedoOperation)()

        Public Sub New(inkCanvas As InkCanvas, strokeService As InkStrokesService)
            _strokeService = strokeService
            AddHandler _strokeService.MoveStrokesEvent, AddressOf StrokeService_MoveStrokesEvent
            AddHandler _strokeService.CutStrokesEvent, AddressOf StrokeService_CutStrokesEvent
            AddHandler _strokeService.PasteStrokesEvent, AddressOf StrokeService_PasteStrokesEvent
            AddHandler inkCanvas.InkPresenter.StrokesCollected, Sub(s, e) AddOperation(New AddStrokeUndoRedoOperation(e.Strokes, _strokeService))
            AddHandler inkCanvas.InkPresenter.StrokesErased, Sub(s, e) AddOperation(New RemoveStrokeUndoRedoOperation(e.Strokes, _strokeService))
        End Sub

        Public Sub Reset()
            undoStack.Clear()
            redoStack.Clear()
        End Sub

        Public ReadOnly Property CanUndo As Boolean
            Get
                Return undoStack.Any()
            End Get
        End Property

        Public ReadOnly Property CanRedo As Boolean
            Get
                Return redoStack.Any()
            End Get
        End Property

        Public Sub Undo()
            If Not CanUndo Then
                Return
            End If

            RemoveHandler _strokeService.MoveStrokesEvent, AddressOf StrokeService_MoveStrokesEvent
            Dim element = undoStack.Pop()
            element.ExecuteUndo()
            redoStack.Push(element)
            AddHandler _strokeService.MoveStrokesEvent, AddressOf StrokeService_MoveStrokesEvent
        End Sub

        Public Sub Redo()
            If Not CanRedo Then
                Return
            End If

            RemoveHandler _strokeService.MoveStrokesEvent, AddressOf StrokeService_MoveStrokesEvent
            Dim element = redoStack.Pop()
            element.ExecuteRedo()
            undoStack.Push(element)
            AddHandler _strokeService.MoveStrokesEvent, AddressOf StrokeService_MoveStrokesEvent
        End Sub

        Public Sub AddOperation(operation As IUndoRedoOperation)
            If operation Is Nothing Then
                Return
            End If

            undoStack.Push(operation)
            redoStack.Clear()
        End Sub

        Private Sub StrokeService_MoveStrokesEvent(sender As Object, e As MoveStrokesEventArgs)
            Dim operation = New MoveStrokesUndoRedoOperation(e.Strokes, e.StartPosition, e.EndPosition, _strokeService)
            AddOperation(operation)
        End Sub

        Private Sub StrokeService_CutStrokesEvent(sender As Object, e As CopyPasteStrokesEventArgs)
            Dim operation = New RemoveStrokeUndoRedoOperation(e.Strokes, _strokeService)
            AddOperation(operation)
        End Sub

        Private Sub StrokeService_PasteStrokesEvent(sender As Object, e As CopyPasteStrokesEventArgs)
            Dim operation = New AddStrokeUndoRedoOperation(e.Strokes, _strokeService)
            AddOperation(operation)
        End Sub
    End Class
End Namespace

File: templates/Uwp/_composition/_shared/Page.Ink.AddUndoRedoService/Services/Ink/InkUndoRedoService.cs
using System.Collections.Generic;
using System.Linq;
using Windows.UI.Xaml.Controls;
using Param_ItemNamespace.EventHandlers.Ink;
using Param_ItemNamespace.Services.Ink.UndoRedo;

namespace Param_ItemNamespace.Services.Ink
{
    public class InkUndoRedoService
    {
        private readonly InkStrokesService _strokeService;

        private Stack<IUndoRedoOperation> undoStack = new Stack<IUndoRedoOperation>();
        private Stack<IUndoRedoOperation> redoStack = new Stack<IUndoRedoOperation>();

        public InkUndoRedoService(InkCanvas inkCanvas, InkStrokesService strokeService)
        {
            _strokeService = strokeService;
            _strokeService.MoveStrokesEvent += StrokeService_MoveStrokesEvent;
            _strokeService.CutStrokesEvent += StrokeService_CutStrokesEvent;
            _strokeService.PasteStrokesEvent += StrokeService_PasteStrokesEvent;

            inkCanvas.InkPresenter.StrokesCollected += (s, e) => AddOperation(new AddStrokeUndoRedoOperation(e.Strokes, _strokeService));
            inkCanvas.InkPresenter.StrokesErased += (s, e) => AddOperation(new RemoveStrokeUndoRedoOperation(e.Strokes, _strokeService));
        }

        public void Reset()
        {
            undoStack.Clear();
            redoStack.Clear();
        }

        public bool CanUndo => undoStack.Any();

        public bool CanRedo => redoStack.Any();

        public void Undo()
        {
            if (!CanUndo)
            {
                return;
            }

            _strokeService.MoveStrokesEvent -= StrokeService_MoveStrokesEvent;

            var element = undoStack.Pop();
            element.ExecuteUndo();
            redoStack.Push(element);

            _strokeService.MoveStrokesEvent += StrokeService_MoveStrokesEvent;
        }

        public void Redo()
        {
            if (!CanRedo)
            {
                return;
            }

            _strokeService.MoveStrokesEvent -= StrokeService_MoveStrokesEvent;

            var element = redoStack.Pop();
            element.ExecuteRedo();
            undoStack.Push(element);

            _strokeService.MoveStrokesEvent += StrokeService_MoveStrokesEvent;
        }

        public void AddOperation(IUndoRedoOperation operation)
        {
            if (operation == null)
            {
                return;
            }

            undoStack.Push(operation);
            redoStack.Clear();
        }

        private void StrokeService_MoveStrokesEvent(object sender, MoveStrokesEventArgs e)
        {
            var operation = new MoveStrokesUndoRedoOperation(e.Strokes, e.StartPosition, e.EndPosition, _strokeService);
            AddOperation(operation);
        }

        private void StrokeService_CutStrokesEvent(object sender, CopyPasteStrokesEventArgs e)
        {
            var operation = new RemoveStrokeUndoRedoOperation(e.Strokes, _strokeService);
            AddOperation(operation);
        }

        private void StrokeService_PasteStrokesEvent(object sender, CopyPasteStrokesEventArgs e)
        {
            var operation = new AddStrokeUndoRedoOperation(e.Strokes, _strokeService);
            AddOperation(operation);
        }
    }
}